## Data Analysis

In [1]:
#import statements
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
import matplotlib.animation as anim
from scipy.stats import gaussian_kde
import numpy as np

import SaturnMoonLibrary as sml
import SaturnMoonLibrary.SaturnMoonSimulation as sms
import SaturnMoonLibrary.SaturnMoonAnalysis as sma

%matplotlib qt

### Loading datasets

In [90]:
# Example usage
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-06_17-14-48.bin' # good for 2d and 3d
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_16-54-17.bin'# good for centered plots #small: 2025-01-11_00-13-15, big: 2025-01-11_16-54-17
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-13_10-42-45.bin ' # good for falling z plot
filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-02-04_10-51-24.bin ' # Same but without ring potential
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_00-13-15.bin' #good for waves
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-12_22-23-34.bin'
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-13_08-41-15.bin'
# filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-14_21-26-21.bin'
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-20_10-57-45.bin'
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-02-04_12-11-31.bin'
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-02-04_12-52-12.bin'
#filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-02-04_22-17-59.bin'
# filepath = 'SaturnModelDatabase/horizons_long_format_data/horizons_long 2025-02-05_13-43-17.bin '
#2025-01-13_11-46-03
dataset = sma.Dataset(filepath)


['Number of Test Particles', 'Inner Radius', 'Outer Radius', 'J2', 'Ring Folder', 'Number of Radial Bins', 'Number of Azimuthal Bins', 'Theta Max', 'Include Shear Forces', 'Include Particle-Moon Collisions', 'Initialisation Method', 'Runtime'] set to standard value
[nan nan nan ... nan nan nan]


### Animations

In [100]:
dataset=datasets[0]
anim_manager = sma.AnimationManager(dataset)
dataset
# 2D Animation
# anim_manager.plot_2d(n_farthest_filter = 5)

# 3D Animation
anim_manager.plot_3d(interval=100)

# Centered Animation
# For a plot centered around "Daphnis" with a square bounds of 1e8 meters
#anim_manager.plot_centered(moon_name="Pan", width=1e7, trail_length=100, interval=1) # To not get dissy one best uses something with low dt and 1 save modularity

# 3D version of Centered
#anim_manager.plot_centered_3d(moon_name="Daphnis", width=4e7, trail_length=100, interval=1) # To not get dissy one best uses something with low dt and 1 save modularity

# Polar plot
#anim_manager.plot_polar_cartesian(r_max =1.5e8, r_min = 0.7e8,theta_max=np.pi*2 ,frame_time=10, trail_length=100, interval=1, heatmap=False)

# Cylindrical plot
# anim_manager.plot_polar_cartesian_with_z(r_max =1.1e8, r_min = 0, z_max=1e5 , z_min=-1e5,elevation=20,azimuth=90, frame_time=10, trail_length=100, interval=1)

In [101]:
dataset.header

{'Moon Names': 'Saturn, Mimas, Enceladus, Tethys, Dione, Rhea, Titan, Hyperion, Iapetus, Phoebe, Janus, Epimetheus, Helene, Telesto, Calypso, Atlas, Prometheus, Pandora, Pan, Daphnis',
 'Moon Count': 20,
 'Initial Data Folder': 'initial_data_moon_count_20_start_date_2017-11-18_00-00-00-000_creation_date_2025-02-04_21-24-42',
 'Epoch': '[2458075.5]',
 'dt': 600,
 'Timesteps': 960000,
 'Number of Test Particles': 10,
 'Saved Points Modularity': 32,
 'Skipped Timesteps': 0,
 'Inner Radius': 70000000.0,
 'Outer Radius': 140000000.0,
 'J2': -0.01629,
 'Ring Folder': 'ring_data_2025-01-09',
 'Number of Radial Bins': 0,
 'Number of Azimuthal Bins': 0,
 'Theta Max': 6.283185307179586,
 'Include Shear Forces': 'False',
 'Include Particle-Moon Collisions': 'False',
 'Numerical Integrator': 'Leapfrog',
 'Initialisation Method': 'Linear',
 'Runtime': 15.496}

---------------------------------------------------------------------------------------------------------
File Name: simulation 2025-01-06_17-14-48.bin 
Creation Date: 2025-01-06 17:14:48
Moon Names: Saturn, Mimas, Enceladus, Tethys, Dione, Rhea, Titan, Hyperion, Iapetus, Phoebe, Janus, Epimetheus, Helene, Telesto, Calypso, Atlas, Prometheus, Pandora, Pan, Daphnis
Moon Count: 20
Initial Data Folder: initial_data_moon_count_20_start_date_2017-11-18_00-00-00-000_creation_date_2024-12-24_21-27-40
Epoch: [2458075.5]
dt: 600.0
Timesteps: 10000
Number of Test Particles: 10000
Saved Points Modularity: 20
Skipped Timesteps: 0
Inner Radius: 70000000.0
Outer Radius: 140000000.0
Include Particle-Moon Collisions: True
---------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------
File Name: simulation 2025-01-11_16-54-17.bin 
Creation Date: 2025-01-11 16:54:17
Moon Names: Saturn, Mimas, Enceladus, Tethys, Dione, Rhea, Titan, Hyperion, Iapetus, Phoebe, Janus, Epimetheus, Helene, Telesto, Calypso, Atlas, Prometheus, Pandora, Pan, Daphnis
Moon Count: 20
Initial Data Folder: initial_data_moon_count_20_start_date_2017-11-18_00-00-00-000_creation_date_2024-12-24_21-27-40
Epoch: [2458075.5]
dt: 600.0
Timesteps: 4000
Number of Test Particles: 100000
Saved Points Modularity: 1
Skipped Timesteps: 3000
Inner Radius: 70000000.0
Outer Radius: 140000000.0
Ring Folder: ring_data_2025-01-09
Include Particle-Moon Collisions: True
Numerical Integrator: Leapfrog
---------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------
File Name: simulation 2025-01-13_08-41-15.bin 
Creation Date: 2025-01-13 08:41:15
Moon Names: Saturn, Mimas, Enceladus, Tethys, Dione, Rhea, Titan, Hyperion, Iapetus, Phoebe, Janus, Epimetheus, Helene, Telesto, Calypso, Atlas, Prometheus, Pandora, Pan, Daphnis
Moon Count: 20
Initial Data Folder: initial_data_moon_count_20_start_date_2017-11-18_00-00-00-000_creation_date_2024-12-24_21-27-40
Epoch: [2458075.5]
dt: 600.0
Timesteps: 100000
Number of Test Particles: 100000
Saved Points Modularity: 1000
Skipped Timesteps: 0
Inner Radius: 116500000.0
Outer Radius: 119000000.0
Ring Folder: ring_data_2025-01-09
Include Particle-Moon Collisions: True
Numerical Integrator: Leapfrog
---------------------------------------------------------------------------------------------------------

### Energy plots

In [2]:
energy=sma.EnergyAnalysis(dataset)
energy.plot_energy(scale="relative")

NameError: name 'dataset' is not defined

In [35]:
t1,t2= 1.3748997502348708,1.3763189283420498

In [39]:
t1*t1/(t1-t2)

-1332.0028780273426

In [ ]:
filenames=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-01.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-17.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-29.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-44.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-54.bin'] # dts increasing in order of 2 leapfrog
# filenames=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-27-53.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-28-37.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-29-06.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-29-24.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-29-39.bin'] #same but Y4
filenames=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-01-03.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-03-25.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-04-56.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-05-50.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-06-32.bin'] #same but Y8
# filenames=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-40-52.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-41-52.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-42-46.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-43-15.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-43-35.bin'] #Y6
datasets= [sma.Dataset(filename) for filename in filenames]
# print("energy")
energies= [sma.EnergyAnalysis(dataset) for dataset in datasets]
plt.figure()

# plt.xlim(0,300)
plots   = [energy.plot_energy(scale="relative",Label="dt="+str(energy.dataset.header["dt"])+"s") for energy in energies]

plt.show()
plt.legend()
# plt.savefig(f"Energy_Y6.pdf")
print([dataset.header["Numerical Integrator"] for dataset in datasets])
print(datasets[-1].positions[-1,:,0])

In [ ]:
filenames1=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-01.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-17.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-29.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-44.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_14-12-54.bin'] # dts increasing in order of 2 leapfrog
filenames2=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-27-53.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-28-37.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-29-06.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-29-24.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_17-29-39.bin'] #same but Y4
filenames3=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-01-03.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-03-25.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-04-56.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-05-50.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-06-32.bin'] #same but Y8
filenames4=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-40-52.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-41-52.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-42-46.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-43-15.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-06_20-43-35.bin'] #Y6
filenames=filenames1+filenames2+filenames3+filenames4
datasets= [sma.Dataset(filename) for filename in filenames]
print("energy")
energies= [sma.EnergyAnalysis(dataset) for dataset in datasets]
plt.figure()
plots   = [energy.plot_energy(scale="absolute",Label=energy.dataset.header["dt"]) for energy in energies]
plt.show()
plt.legend()

In [3]:
datasets[0].positions.shape

(30001, 20, 6)

In [130]:
dat_index1=1
dat_index2=2
moon_index=1
dt=datasets[0].header["dt"]*(datasets[0].header["Saved Points Modularity"])
fig, (axes)= plt.subplots(5,5)
for dat_index1 in range(0,5):
    for dat_index2 in range(0,5):
        axes[dat_index1][dat_index2].plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),np.sqrt(np.sum((datasets[dat_index1].positions[:,moon_index,:3]-datasets[dat_index2].positions[:,moon_index,:3])**2,axis=-1)),label=datasets[dat_index2].header["dt"])
        # axes[dat_index1][dat_index2].set_xlim(0,400)
        axes[dat_index1][dat_index2].set_title(f"{dat_index2}-{dat_index1}")
plt.legend()

In [9]:
dat_index1=1
dat_index2=2
moon_index=2
fig, (axes)= plt.subplots(5,1)
fig.suptitle(datasets[dat_index1].moons[moon_index].name)

for dat_index1 in range(0,5):
    axes[dat_index1].plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),datasets[dat_index1].positions[:,moon_index,0])

In [183]:
datasets[dat_index2].header

{'Moon Names': 'Saturn, Mimas, Enceladus, Tethys, Dione, Rhea, Titan, Hyperion, Iapetus, Phoebe, Janus, Epimetheus, Helene, Telesto, Calypso, Atlas, Prometheus, Pandora, Pan, Daphnis',
 'Moon Count': 20,
 'Initial Data Folder': 'initial_data_moon_count_20_start_date_2017-11-18_00-00-00-000_creation_date_2025-02-04_21-24-42',
 'Epoch': '[2458075.5]',
 'dt': 1200,
 'Timesteps': 480000,
 'Number of Test Particles': 0,
 'Saved Points Modularity': 16,
 'Skipped Timesteps': 0,
 'Inner Radius': 70000000.0,
 'Outer Radius': 140000000.0,
 'J2': -0.01629,
 'Ring Folder': 'ring_data_2025-01-09',
 'Number of Radial Bins': 0,
 'Number of Azimuthal Bins': 0,
 'Theta Max': 6.283185307179586,
 'Include Shear Forces': 'False',
 'Include Particle-Moon Collisions': 'False',
 'Numerical Integrator': 'Yoshida 8th Order',
 'Initialisation Method': 'linear',
 'Runtime': 91.208}

In [197]:
dat_index1=0
dat_index2=4
moon_index=2
plt.figure()
dimension_index=0
dt=datasets[0].header["dt"]*(datasets[0].header["Saved Points Modularity"])
datasets[dat_index1].moons[moon_index].calculate_orbital_elements()[0]
a=np.average(datasets[0].moons[moon_index].calculate_orbital_elements()[0]['semimajor_axis'])
plt.plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),(datasets[dat_index1].positions[:,moon_index,dimension_index]-datasets[dat_index2].positions[:,moon_index,dimension_index])/a)
plt.ylabel("relative difference in x coordinate")
plt.xlabel("time (days)")
# plt.xlim(0,100)
plt.show()
plt.savefig(f"{datasets[dat_index1].moons[moon_index].name}oscillations_{datasets[dat_index1].header['dt']}and{datasets[dat_index2].header['dt']}Y6.pdf")
print(datasets[dat_index1].moons[moon_index].name)

Enceladus


In [8]:
plt.plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),(datasets[dat_index1].positions[:,moon_index,0]))
plt.plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),datasets[dat_index2].positions[:,moon_index,0])

In [11]:
data=datasets[dat_index2].positions[:,moon_index,1]
N = len(data)
freqs = np.fft.fftfreq(N, dt)  # Frequency bins
fft_values = np.fft.fft(data)   # FFT of the data
fft_values = fft_values / len(data)
plt.plot(freqs, np.abs(fft_values))  # Only return the magnitude
1/freqs[np.argmax(np.abs(fft_values))]/60/60/24

0.5666232270005855

In [83]:
1/((1.4921416492589274-1.3763189283420498)/1.3763189283420498)*1.3763189283420498

16.354768542107173

In [ ]:
data=datasets[dat_index1].positions[:,moon_index,0]
N = len(data)
freqs = np.fft.fftfreq(N, dt)  # Frequency bins
fft_values = np.fft.fft(data)   # FFT of the data
fft_values = fft_values / len(data)
plt.plot(freqs, np.abs(fft_values))  # Only return the magnitude
1/freqs[np.argmax(np.abs(fft_values))]/60/60/24

600


In [ ]:
dat_index1=1
dat_index2=2
moon_index=2
plt.title(datasets[dat_index1].moons[moon_index].name)
for dat_index1 in range(4,-1,-1):
    plt.plot((datasets[dat_index1].moons[moon_index].calculate_orbital_elements()[0]['period'])/60/60/24)
    print(np.average(datasets[dat_index1].moons[moon_index].calculate_orbital_elements()[0]['period'])/60/60/24)

1.4153353232844994
1.378832282599597
1.375034263789278
1.3746474619713327
1.3745914369226144


In [55]:
datasets[dat_index1].moons[moon_index].calculate_orbital_elements()

({'inclination': array([0.34613677, 0.38094199, 0.44153349, ..., 0.3486895 , 0.39223486,
         0.44894461]),
  'ascending_node_longitude': array([235.43831902, 219.88110909, 211.09909401, ..., 270.24417788,
         260.30979256, 255.94778453]),
  'argument_of_periapsis': array([ 52.42274145,  60.84303277,  67.05380113, ..., 264.75662173,
         274.72671616, 279.24276193]),
  'eccentricity': array([0.00102681, 0.00102106, 0.00102164, ..., 0.00297741, 0.00298068,
         0.00299462]),
  'semimajor_axis': array([5.27221885e+08, 5.27222766e+08, 5.27231024e+08, ...,
         5.27215772e+08, 5.27218516e+08, 5.27228276e+08]),
  'period': array([390546.5451012 , 390547.52385495, 390556.69965818, ...,
         390539.75277984, 390542.80179036, 390553.64598478])},
 {'angular_momentum_vector': array([[-1.93509273e+10,  1.88517760e+10,  4.47183503e+12],
         [-1.51242752e+10,  2.55980494e+10,  4.47182155e+12],
         [-1.35205978e+10,  3.16983984e+10,  4.47182263e+12],
         ...,


### J2 vs no J2 ring vs No ring

In [ ]:
filenames=["simulation 2025-02-07_10-56-50.bin","simulation 2025-02-07_11-02-10.bin","simulation 2025-02-07_10-45-58.bin","simulation 2025-02-07_10-32-18.bin"]
filenames=["simulation 2025-02-07_19-45-27.bin","simulation 2025-02-07_19-52-31.bin","simulation 2025-02-07_19-59-21.bin","simulation 2025-02-07_20-11-12.bin"] #accidently left collisions on and not linearly placed
filepaths=[".\\SaturnModelDatabase\\simulation_data\\"+name for name in filenames]
filepaths=['.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-07_20-36-20.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-07_20-44-17.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-07_20-54-38.bin', '.\\SaturnModelDatabase\\simulation_data\\simulation 2025-02-07_21-04-05.bin']

datasets= [sma.Dataset(filename) for filename in filepaths]
energies= [sma.EnergyAnalysis(dataset) for dataset in datasets]
plt.figure()
plots   = [energy.plot_energy(scale="absolute",Label=str(energy.dataset.header["J2"])+energy.dataset.header["Ring Folder"]) for energy in energies]
plt.show()
plt.legend()

[ 6.77940297e+04  2.85542849e+05  3.74743375e+03 ...  1.25142889e+04
 -1.06103608e+04  2.00971540e+01]
[ 6.77940297e+04  2.85542849e+05  3.74743375e+03 ... -3.21642308e+03
 -1.61218834e+04  1.48563762e+00]
[ 6.77940297e+04  2.85542849e+05  3.74743375e+03 ...  1.55437610e+04
  5.46986284e+03 -1.43685812e+01]
[ 6.77940297e+04  2.85542849e+05  3.74743375e+03 ... -6.53252867e+03
 -1.50162542e+04 -4.63450554e+00]


findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font fam

In [ ]:
fig = plt.figure(figsize=(12, 10))

# Create 2x2 subplots with 3D axes
axes = [fig.add_subplot(2, 2, i + 1) for i in range(4)]

fig.suptitle("Mosaic Plot of Timesteps", fontsize=16)

# Loop over axes and timesteps and call the plot function
plt.sca(axes[0])
energies[0].plot_energy()
plt.title(f"J2={energies[0].dataset.header['J2']},{'Ring' if energies[0].dataset.header['Ring Folder']=='ring_data_2025-01-09' else 'No Ring'}")
plt.sca(axes[1])
energies[1].plot_energy()
plt.title(f"J2={energies[1].dataset.header['J2']},{'Ring' if energies[1].dataset.header['Ring Folder']=='ring_data_2025-01-09' else 'No Ring'}")
plt.sca(axes[2])
energies[2].plot_energy()
plt.title(f"J2={energies[2].dataset.header['J2']},{'Ring' if energies[2].dataset.header['Ring Folder']=='ring_data_2025-01-09' else 'No Ring'}")
plt.sca(axes[3])
energies[3].plot_energy()
plt.title(f"J2={energies[3].dataset.header['J2']},{'Ring' if energies[3].dataset.header['Ring Folder']=='ring_data_2025-01-09' else 'No Ring'}")
plt.tight_layout(rect=[0, 0, 1, 0.95])  # Adjust layout to fit title
plt.show()

findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font fam

findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font family 'Humor Sans' not found.
findfont: Font family 'Comic Neue' not found.
findfont: Font family 'xkcd' not found.
findfont: Font family 'xkcd Script' not found.
findfont: Font fam

In [43]:
datasets[dataset_index2].moons[0].calculate_orbital_elements()[0].keys()

NameError: name 'dataset_index2' is not defined

In [49]:
for i in range(4):
    print((i,datasets[i].header["J2"],datasets[i].header["Ring Folder"]))

(0, 0, 'ring_data_2025-01-08')
(1, 0, 'ring_data_2025-01-09')
(2, -0.01629, 'ring_data_2025-01-08')
(3, -0.01629, 'ring_data_2025-01-09')


In [113]:
dataset_index1=1
dataset_index2=3
orbital_elements=['inclination', 'ascending_node_longitude', 'argument_of_periapsis', 'eccentricity', 'semimajor_axis', 'period']
orbital_element=orbital_elements[0]
a1=[(datasets[dataset_index1].moons[moon_index].name,np.average(datasets[dataset_index1].moons[moon_index].calculate_orbital_elements()[0][orbital_element]),np.std(datasets[dataset_index1].moons[moon_index].calculate_orbital_elements()[0][orbital_element])) for moon_index in range(1,20)]
a2=[(datasets[dataset_index2].moons[moon_index].name,np.average(datasets[dataset_index2].moons[moon_index].calculate_orbital_elements()[0][orbital_element]),np.std(datasets[dataset_index2].moons[moon_index].calculate_orbital_elements()[0][orbital_element])) for moon_index in range(1,20)]
difference=[(el1[0],el1[1]-el2[1],el1[2]-el2[2]) for el1,el2 in zip(a1,a2)]
rel_dif   =[(el[0],el[1]/el1[1],el[2]) for el,el1 in zip(difference,a1)]
print(a1)
print(a2)
print(difference)
print(rel_dif)

[('Mimas', 1.353131740730694, 0.16253070360424235), ('Enceladus', 0.010441237910425534, 0.0012548086233447834), ('Tethys', 1.3014599545573462, 0.15625468574176632), ('Dione', 0.02985455052386695, 0.0036281102956866864), ('Rhea', 0.420512686506867, 0.050504425431949995), ('Titan', 0.4263592931138565, 0.05124613040127528), ('Hyperion', 1.0720074237049126, 0.13060722185981924), ('Iapetus', 16.923443925430195, 1.9225436434744452), ('Phoebe', 148.53210464177212, 3.2150009168248572), ('Janus', 0.14499971520944924, 0.017416473726084684), ('Epimetheus', 0.31496953451107773, 0.03783377771941416), ('Helene', 0.22936455949402826, 0.027554427788919096), ('Telesto', 1.0441228961492046, 0.12537266194782362), ('Calypso', 1.5172369574729048, 0.18213926277409656), ('Atlas', 0.0028217801277123466, 0.000338974530531272), ('Prometheus', 0.00903659838864572, 0.001085320805282968), ('Pandora', 0.04395926946682318, 0.005279617309368759), ('Pan', 1.1415387595024539e-05, 5.620634139670076e-06), ('Daphnis', 0.0

In [114]:
print(orbital_element)
tester=difference.copy()
tester.sort(key=lambda x: abs(x[1]))
tester


inclination


[('Enceladus', 4.694653358595924e-08, -6.910344815900388e-09),
 ('Titan', -6.863639745446903e-08, 8.350671869272741e-09),
 ('Pan', -9.114931271003639e-08, 4.683046552332609e-07),
 ('Phoebe', 1.4505957324217889e-07, -5.3200367222672185e-08),
 ('Rhea', -3.0965083791523185e-07, 5.353229706045459e-08),
 ('Iapetus', -3.354860709237073e-07, -1.292033180444463e-07),
 ('Prometheus', -6.686292345972622e-07, -1.2190823363809888e-07),
 ('Calypso', 8.995315565751838e-07, -1.9869146527240478e-06),
 ('Hyperion', 9.145167843005453e-07, 9.79543944013006e-08),
 ('Janus', 1.5172110500449865e-06, -3.388240047488833e-07),
 ('Epimetheus', 1.8548347818780897e-06, -1.1197492440634393e-06),
 ('Helene', 3.2376865390648657e-06, 1.0644657592179252e-07),
 ('Telesto', 3.6168052603979817e-06, -9.450884327910636e-07),
 ('Tethys', -5.275047057162752e-06, -1.732453410091539e-06),
 ('Dione', -5.683291034671439e-06, -1.2030041855342656e-06),
 ('Pandora', -6.409622414597915e-06, -9.661128536793406e-07),
 ('Mimas', 9.5662

In [115]:
a1=[(datasets[dataset_index1].test_particles[moon_index].name,np.average(datasets[dataset_index1].test_particles[moon_index].calculate_orbital_elements()[0][orbital_element]),np.std(datasets[dataset_index1].test_particles[moon_index].calculate_orbital_elements()[0][orbital_element])) for moon_index in range(0,10)]
a2=[(datasets[dataset_index2].test_particles[moon_index].name,np.average(datasets[dataset_index2].test_particles[moon_index].calculate_orbital_elements()[0][orbital_element]),np.std(datasets[dataset_index2].test_particles[moon_index].calculate_orbital_elements()[0][orbital_element])) for moon_index in range(0,10)]
difference=[(el1[0],el1[1]-el2[1],el1[2]-el2[2]) for el1,el2 in zip(a1,a2)]
rel_dif   =[(el[0],el[1]/el1[1],el[2]) for el,el1 in zip(difference,a1)]
print(a1)
print(a2)
print(difference)
print(rel_dif)

[('0', 0.003751122775594151, 0.00045051607024384615), ('1', 0.0033743582947315114, 0.00040526155113160896), ('2', 0.003066872366889308, 0.0003683313087966505), ('3', 0.0028109513588278593, 0.0003376009249119524), ('4', 0.00259389251322165, 0.0003115260390654734), ('5', 0.0024076778199520715, 0.0002891608113919653), ('6', 0.0022471544924523848, 0.00026988141975596374), ('7', 0.0021066828347742358, 0.00025301717809023453), ('8', 0.0019821399301783327, 0.00023805470325200431), ('9', 0.003932075057576427, 0.0009405385139971428)]
[('0', 0.0037444041727252133, 0.0004497684970661927), ('1', 0.00336948039794842, 0.0004047268494087352), ('2', 0.003062623542560573, 0.00036785432623303704), ('3', 0.002806128574994608, 0.00033704543874598356), ('4', 0.0025904202078481566, 0.0003111310860696568), ('5', 0.002405161436693647, 0.00028887679779177896), ('6', 0.0022446551839457535, 0.00026959386710129), ('7', 0.0021043792545845075, 0.0002527494363446333), ('8', 0.0019811347429635356, 0.00023794264660121

In [116]:
print(orbital_element)
tester=difference.copy()
tester.sort(key=lambda x: abs(x[1]))
tester

inclination


[('8', 1.0051872147971534e-06, 1.1205665079064113e-07),
 ('7', 2.3035801897282326e-06, 2.677417456012224e-07),
 ('6', 2.4993085066312565e-06, 2.8755265467372577e-07),
 ('5', 2.5163832584247127e-06, 2.8401360018632475e-07),
 ('4', 3.472305373493443e-06, 3.949529958166163e-07),
 ('2', 4.248824328735145e-06, 4.769825636134393e-07),
 ('3', 4.822783833251156e-06, 5.554861659688216e-07),
 ('1', 4.877896783091282e-06, 5.347017228737698e-07),
 ('0', 6.718602868937672e-06, 7.475731776534502e-07),
 ('9', -0.022500823500817533, -0.012401471349173576)]

In [105]:
raise ValueError

ValueError: 

In [ ]:
moon_index=8
orbital_element=orbital_elements[5]
dt=datasets[0].header["dt"]*(datasets[0].header["Saved Points Modularity"])
plt.plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),datasets[dataset_index1].moons[moon_index].calculate_orbital_elements()[0][orbital_element])
plt.plot(dt/60/60/24*np.arange(0,len(datasets[0].positions)),datasets[dataset_index2].moons[moon_index].calculate_orbital_elements()[0][orbital_element])

In [ ]:
data=datasets[dataset_index1].moons[moon_index].calculate_orbital_elements()[0][orbital_element]
N = len(data)
start=1
freqs = np.fft.fftfreq(N, dt)  # Frequency bins
fft_values = np.fft.fft(data)   # FFT of the data
fft_values = fft_values / len(data)
plt.plot(freqs[start:], np.abs(fft_values)[start:])  # Only return the magnitude
1/freqs[np.argmax(np.abs(fft_values)[start:])]/60/60/24

672.0486111111112

array([0.00000000e+00, 5.55550926e-10, 1.11110185e-09, ...,
       3.33313889e-05, 3.33319445e-05, 3.33325000e-05])

In [60]:
1/700/60/60/24

1.6534391534391535e-08

### Collision Analysis

In [11]:
collisions_ls=[sma.CollisionAnalysis(dataset) for dataset in datasets]
[collisions.collisions_count() for collisions in collisions_ls]

[None, None, None, None, None]

In [18]:
collisions_ls[4].collisions_count()

In [29]:
np.nonzero(np.abs(datasets[3].positions[-1,:,0])>1e11)

(array([20], dtype=int64),)

In [36]:
for i in range(20,30):
    plt.plot(datasets[4].positions[:,i,0])

c:\Users\kdvan\AppData\Local\Programs\Python\Python311\Lib\site-packages\ipykernel\eventloops.py:113: UserWarning: Attempting to set identical low and high ylims makes transformation singular; automatically expanding.
  app.exec_()
c:\Users\kdvan\AppData\Local\Programs\Python\Python311\Lib\site-packages\ipykernel\eventloops.py:113: UserWarning: Attempting to set identical low and high xlims makes transformation singular; automatically expanding.
  app.exec_()


In [30]:
np.sum(np.abs(datasets[4].positions[:,:,0])==1e12,axis=1)

array([0, 0, 0, ..., 0, 0, 0])

### Wave Analysis

In [12]:
# Wavelengths
wave_analysis = sma.WaveAnalysis(dataset, bins=20, plot_type="2d")

# Perform analysis at timestep 0
wave_analysis.azimuthal_bin_analysis(i=50)

# Compute wavelengths in a specific range
result = wave_analysis.compute_wavelength_2d(i=0, r_min=1.1e8, r_max=1.4e8)
print(result)

# Plot the wavelength over time
wave_analysis.plot_wavelength_over_time(r_min=1.1e8, r_max=1.4e8)


{'average_wavelength': 8925387.8072236, 'medianwavelength': 7268028.108811073, 'num_peaks': 87}


In [4]:
# Animations
wave_analysis = sma.WaveAnalysis(dataset, bins=20, plot_type="2d")

# Perform analysis at timestep 0
wave_analysis.azimuthal_bin_analysis(i=0)

# Animate and ensure it runs smoothly
animation_2d, animation_3d = wave_analysis.animate()


archive:

In [5]:
import numpy as np
import matplotlib.pyplot as plt

class FourierRings:
    def __init__(self, dataset):
        """
        Initializes the FourierRings class with a given dataset.

        Args:
            dataset (Dataset): The dataset object containing simulation data.
        """
        self.dataset = dataset
        self.time_steps = self.dataset.positions.shape[0]
        self.num_test_particles = self.dataset.num_test_particles

    def cartesian_to_polar(self, x, y):
        """
        Converts Cartesian coordinates (x, y) to polar (r, theta).

        Args:
            x (ndarray): x-coordinate values.
            y (ndarray): y-coordinate values.

        Returns:
            tuple: Polar coordinates (r, theta).
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, np.mod(theta, 2 * np.pi)  # Ensure theta is in the range [0, 2pi]

    def compute_fft(self, data, dt=1):
        """
        Computes the FFT of a dataset.

        Args:
            data (ndarray): The data to apply FFT on.
            dt (float): Time step between data points, default is 1.

        Returns:
            tuple: Frequencies and corresponding FFT values.
        """
        N = len(data)
        freqs = np.fft.fftfreq(N, dt)  # Frequency bins
        fft_values = np.fft.fft(data)   # FFT of the data
        fft_values = fft_values / len(data)
        return freqs, np.abs(fft_values)  # Only return the magnitude

    def analyze_ring_waveform(self):
        """
        Analyzes the waveforms in the test particles' radial distances using FFT.

        Returns:
            None: Displays a plot of the FFT spectrum.
        """
        # Extract radial distances of all test particles across all time steps
        radial_distances = np.zeros((self.time_steps, self.num_test_particles))
        
        for i in range(self.num_test_particles):
            x = self.dataset.positions[:, self.dataset.num_moons + i, 0]
            y = self.dataset.positions[:, self.dataset.num_moons + i, 1]
            radial_distances[:, i] = self.cartesian_to_polar(x, y)[0]

        # Compute the FFT for each test particle's radial distance
        plt.figure(figsize=(10, 6))
        for i in range(self.num_test_particles):
            radial_data = radial_distances[:, i]
            freqs, fft_values = self.compute_fft(radial_data)

            # Plot the FFT results for the test particle
            plt.plot(freqs, fft_values, label=f"Test Particle {i + 1}")

        # Set plot labels and title
        plt.xlabel('Frequency [Hz]')
        plt.ylabel('Amplitude')
        plt.title('FFT of Test Particles\' Radial Distance')
        #plt.legend() not safe at all
        plt.xlim(0.00001, 0.01)
        print(self.dataset.header['dt'])
        plt.show()

# Example usage
filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_00-13-15.bin'
dataset = Dataset(filepath)
fourier_rings = FourierRings(dataset)

# Analyze the waveform in the ring using FFT
fourier_rings.analyze_ring_waveform()


NameError: name 'Dataset' is not defined

In [12]:
import numpy as np
import matplotlib.pyplot as plt

class FourierRings:
    def __init__(self, dataset):
        """
        Initializes the FourierRings class with a given dataset.

        Args:
            dataset (Dataset): The dataset object containing simulation data.
        """
        self.dataset = dataset
        self.time_steps = self.dataset.positions.shape[0]
        self.num_test_particles = self.dataset.num_test_particles

    def cartesian_to_polar(self, x, y):
        """
        Converts Cartesian coordinates (x, y) to polar (r, theta).

        Args:
            x (ndarray): x-coordinate values.
            y (ndarray): y-coordinate values.

        Returns:
            tuple: Polar coordinates (r, theta).
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, np.mod(theta, 2 * np.pi)  # Ensure theta is in the range [0, 2pi]

    def compute_fft(self, data, dt=1):
        """
        Computes the FFT of a dataset.

        Args:
            data (ndarray): The data to apply FFT on.
            dt (float): Time step between data points, default is 1.

        Returns:
            tuple: Frequencies and corresponding FFT values.
        """
        N = len(data)
        freqs = np.fft.fftfreq(N, dt)  # Frequency bins
        fft_values = np.fft.fft(data)   # FFT of the data
        fft_values = fft_values / len(data)
        return freqs[:N // 2], np.abs(fft_values[:N // 2])  # Only return positive frequencies

    def group_particles(self, radial_distances, num_groups=5):
        """
        Groups test particles based on their mean radial distances.

        Args:
            radial_distances (ndarray): Radial distances of all test particles over time.
            num_groups (int): Number of groups to divide particles into.

        Returns:
            dict: A dictionary where keys are group indices and values are particle indices.
        """
        mean_distances = np.mean(radial_distances, axis=0)
        sorted_indices = np.argsort(mean_distances)
        group_size = self.num_test_particles // num_groups
        
        groups = {}
        for group_idx in range(num_groups):
            start_idx = group_idx * group_size
            end_idx = (group_idx + 1) * group_size
            groups[group_idx] = sorted_indices[start_idx:end_idx]
        
        return groups

    def analyze_ring_waveform_by_group(self, num_groups=5):
        """
        Analyzes the waveforms of radial distances for groups of test particles using FFT.

        Args:
            num_groups (int): Number of groups to divide particles into.

        Returns:
            None: Displays a plot of the FFT spectrum for each group.
        """
        # Extract radial distances of all test particles across all time steps
        radial_distances = np.zeros((self.time_steps, self.num_test_particles))
        
        for i in range(self.num_test_particles):
            x = self.dataset.positions[:, self.dataset.num_moons + i, 0]
            y = self.dataset.positions[:, self.dataset.num_moons + i, 1]
            radial_distances[:, i] = self.cartesian_to_polar(x, y)[0]

        # Group particles based on their mean radial distances
        groups = self.group_particles(radial_distances, num_groups=num_groups)

        plt.figure(figsize=(12, 8))
        for group_idx, particle_indices in groups.items():
            # Compute the mean radial distance for the group
            group_data = np.mean(radial_distances[:, particle_indices], axis=1)
            freqs, fft_values = self.compute_fft(group_data, dt=self.dataset.header['dt'])

            # Plot the FFT for the group
            plt.plot(freqs, fft_values, label=f"Group {group_idx + 1} ({len(particle_indices)} particles)")

        # Set plot labels and title
        plt.xlabel('Frequency [Hz]')
        plt.ylabel('Amplitude')
        plt.title('FFT of Test Particles\' Radial Distance by Group')
        plt.legend()
        plt.xlim(0.00001, 1/(2*((self.dataset.header['dt']))))
        plt.show()

# Example usage
filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_00-13-15.bin'
dataset = Dataset(filepath)
fourier_rings = FourierRings(dataset)

# Analyze the waveform in the ring using FFT by groups
fourier_rings.analyze_ring_waveform_by_group(num_groups=5)


In [18]:
import numpy as np
import matplotlib.pyplot as plt

class FourierRings:
    def __init__(self, dataset, num_bins=1000):
        """
        Initializes the FourierRings class with a given dataset and radial bins.

        Args:
            dataset (Dataset): The dataset object containing simulation data.
            num_bins (int): Number of radial bins to compute the density.
        """
        self.dataset = dataset
        self.num_bins = num_bins
        self.time_steps = self.dataset.positions.shape[0]
        self.num_test_particles = self.dataset.num_test_particles

    def cartesian_to_polar(self, x, y):
        """
        Converts Cartesian coordinates (x, y) to polar (r, theta).

        Args:
            x (ndarray): x-coordinate values.
            y (ndarray): y-coordinate values.

        Returns:
            tuple: Polar coordinates (r, theta).
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, np.mod(theta, 2 * np.pi)

    def compute_fft(self, data, dt=1):
        """
        Computes the FFT of a dataset.

        Args:
            data (ndarray): The data to apply FFT on.
            dt (float): Time step between data points, default is 1.

        Returns:
            tuple: Frequencies and corresponding FFT values.
        """
        N = len(data)
        freqs = np.fft.fftfreq(N, d=dt)  # Frequency bins
        fft_values = np.fft.fft(data)    # FFT of the data
        fft_values = fft_values / N      # Normalize the FFT amplitude
        return freqs[:N // 2], np.abs(fft_values[:N // 2])  # Return positive frequencies

    def compute_density(self):
        """
        Computes the density profile of test particles in radial bins over time.

        Returns:
            ndarray: Radial density as a 2D array (time_steps x num_bins).
        """
        # Radial bin edges
        max_radius = 1.5  # Adjust based on the system's scale
        min_radius = 0.5  # Adjust based on the system's scale
        bin_edges = np.linspace(min_radius, max_radius, self.num_bins + 1)
        
        # Initialize density array
        density = np.zeros((self.time_steps, self.num_bins))
        
        for t in range(self.time_steps):
            x = self.dataset.positions[t, self.dataset.num_moons:, 0]
            y = self.dataset.positions[t, self.dataset.num_moons:, 1]
            r = self.cartesian_to_polar(x, y)[0]
            
            # Compute histogram for this time step
            density[t, :], _ = np.histogram(r, bins=bin_edges)
        
        return density, bin_edges

    def analyze_density_waveform(self):
        """
        Analyzes the density waveform by computing FFT of density in each radial bin.

        Returns:
            None: Displays a plot of the FFT spectrum for each radial bin.
        """
        # Compute radial density
        density, bin_edges = self.compute_density()
        bin_centers = 0.5 * (bin_edges[:-1] + bin_edges[1:])
        
        plt.figure(figsize=(12, 8))
        
        # Compute FFT for each radial bin
        for i in range(self.num_bins):
            density_time_series = density[:, i]
            freqs, fft_values = self.compute_fft(density_time_series, dt=self.dataset.header['dt'])
            
            # Plot the FFT result
            plt.plot(freqs, fft_values, label=f"Radial Bin {i + 1} ({bin_centers[i]:.2f} AU)")
        
        # Set plot labels and title
        plt.xlabel('Frequency [Hz]')
        plt.ylabel('Amplitude')
        plt.title('FFT of Density Waveform in Radial Bins')
        plt.legend(loc="upper right", fontsize='small', ncol=2)
        plt.xlim(0, 1/(2*((self.dataset.header['dt']))))
        plt.show()

# Example usage
filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_16-54-17.bin'
dataset = Dataset(filepath)
fourier_rings = FourierRings(dataset)

# Analyze the density waveform using FFT
fourier_rings.analyze_density_waveform()


In [19]:
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation

class FourierRings:
    def __init__(self, dataset):
        """
        Initializes the FourierRings class with a given dataset.

        Args:
            dataset (Dataset): The dataset object containing simulation data.
        """
        self.dataset = dataset
        self.time_steps = self.dataset.positions.shape[0]
        self.num_test_particles = self.dataset.num_test_particles

    def cartesian_to_polar(self, x, y):
        """
        Converts Cartesian coordinates (x, y) to polar (r, theta).
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, np.mod(theta, 2 * np.pi)  # Ensure theta is in the range [0, 2pi]

    def compute_density(self, num_bins):
        """
        Computes the density in radial bins for each time step.

        Args:
            num_bins (int): Number of radial bins.

        Returns:
            ndarray: 2D array where each row corresponds to density in radial bins at a time step.
        """
        radial_distances = np.zeros((self.time_steps, self.num_test_particles))
        for i in range(self.num_test_particles):
            x = self.dataset.positions[:, self.dataset.num_moons + i, 0]
            y = self.dataset.positions[:, self.dataset.num_moons + i, 1]
            radial_distances[:, i] = self.cartesian_to_polar(x, y)[0]

        # Define radial bins
        r_min, r_max = np.min(radial_distances), np.max(radial_distances)
        radial_bins = np.linspace(r_min, r_max, num_bins + 1)

        # Compute density for each time step
        density = np.zeros((self.time_steps, num_bins))
        for t in range(self.time_steps):
            density[t], _ = np.histogram(radial_distances[t], bins=radial_bins, density=True)

        return density, radial_bins

    def animate_density(self, num_bins=50):
        """
        Animates the density in a polar plot.

        Args:
            num_bins (int): Number of radial bins for density calculation.
        """
        # Compute density
        density, radial_bins = self.compute_density(num_bins)

        # Create theta for polar plot
        theta = np.linspace(0, 2 * np.pi, num_bins, endpoint=False)

        # Set up the polar plot
        fig, ax = plt.subplots(subplot_kw={'projection': 'polar'})
        line, = ax.plot([], [], lw=2)
        ax.set_ylim(0, np.max(density))  # Set y-axis limit to maximum density

        def update(frame):
            """
            Update function for animation.
            """
            line.set_data(theta, density[frame])
            return line,

        ani = FuncAnimation(
            fig, update, frames=self.time_steps, interval=100, blit=True
        )

        # Show the animation
        plt.show()

# Example usage
filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_00-13-15.bin'
dataset = Dataset(filepath)
fourier_rings = FourierRings(dataset)

# Animate the density waveform
fourier_rings.animate_density(num_bins=50)


In [11]:
from scipy.fft import fft, fftfreq
import matplotlib.pyplot as plt
import numpy as np


class WaveFrequencyAnalyzer:
    """
    Analyzes the frequency components of the motion of moons and test particles in the dataset.

    Attributes:
        dataset (Dataset): The simulation dataset with positional data for moons and test particles.
        dt (float): The time step between simulation points, extracted from the dataset header.
    """
    def __init__(self, dataset):
        self.dataset = dataset
        self.dt = dataset.header['dt']  # Time step between simulation points

    def _cartesian_to_polar(self, x, y):
        """
        Converts Cartesian coordinates (x, y) to polar coordinates (r, theta).

        Args:
            x (ndarray): x-coordinates.
            y (ndarray): y-coordinates.

        Returns:
            tuple: (r, theta), where r is the radial distance and theta is the angle.
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, np.mod(theta, 2 * np.pi)  # Ensure theta is in [0, 2pi]

    def _fft_analysis(self, time_series):
        """
        Performs FFT analysis on a given time series.

        Args:
            time_series (ndarray): The input time series data.

        Returns:
            tuple: Frequencies and FFT result (amplitude).
        """
        # Number of timesteps
        n = len(time_series)
        # Compute the FFT
        fft_values = fft(time_series)
        # Compute the corresponding frequencies
        frequencies = fftfreq(n, d=self.dt)

        # Return frequencies and the magnitude of the FFT
        return frequencies, np.abs(fft_values)

    def analyze_moon(self, moon_name, plot=True):
        """
        Analyzes the frequency components for a specific moon in the dataset.

        Args:
            moon_name (str): Name of the moon to analyze.
            plot (bool): Whether to plot the results. Default is True.

        Returns:
            dict: Dictionary containing FFT results for r, theta, and z.
        """
        moon = next(moon for moon in self.dataset.moons if moon.name == moon_name)

        # Convert Cartesian coordinates to polar (r, theta) and extract z
        r, theta = self._cartesian_to_polar(moon.pos[:, 0], moon.pos[:, 1])
        z = moon.pos[:, 2]

        # Perform FFT for r, theta, and z
        freq_r, fft_r = self._fft_analysis(r)
        freq_theta, fft_theta = self._fft_analysis(theta)
        freq_z, fft_z = self._fft_analysis(z)

        # Optionally plot the FFT results
        if plot:
            self._plot_fft_results(moon_name, freq_r, fft_r, freq_theta, fft_theta, freq_z, fft_z)

        # Return the results as a dictionary
        return {
            'r': (freq_r, fft_r),
            'theta': (freq_theta, fft_theta),
            'z': (freq_z, fft_z)
        }

    def analyze_test_particles(self, particle_index=0, plot=True):
        """
        Analyzes the frequency components for a specific test particle in the dataset.

        Args:
            particle_index (int): Index of the test particle to analyze.
            plot (bool): Whether to plot the results. Default is True.

        Returns:
            dict: Dictionary containing FFT results for r, theta, and z.
        """
        test_particle_pos = self.dataset.test_particles[particle_index].pos

        # Convert Cartesian coordinates to polar (r, theta) and extract z
        r, theta = self._cartesian_to_polar(test_particle_pos[:, 0], test_particle_pos[:, 1])
        z = test_particle_pos[:, 2]

        # Perform FFT for r, theta, and z
        freq_r, fft_r = self._fft_analysis(r)
        freq_theta, fft_theta = self._fft_analysis(theta)
        freq_z, fft_z = self._fft_analysis(z)

        # Optionally plot the FFT results
        if plot:
            self._plot_fft_results(f"Test Particle {particle_index}", freq_r, fft_r, freq_theta, fft_theta, freq_z, fft_z)

        # Return the results as a dictionary
        return {
            'r': (freq_r, fft_r),
            'theta': (freq_theta, fft_theta),
            'z': (freq_z, fft_z)
        }

    def _plot_fft_results(self, name, freq_r, fft_r, freq_theta, fft_theta, freq_z, fft_z):
        """
        Plots the FFT results for the radial, angular, and z components.

        Args:
            name (str): Name of the body being analyzed.
            freq_r (ndarray): Frequencies for the radial component.
            fft_r (ndarray): FFT amplitude for the radial component.
            freq_theta (ndarray): Frequencies for the angular component.
            fft_theta (ndarray): FFT amplitude for the angular component.
            freq_z (ndarray): Frequencies for the z component.
            fft_z (ndarray): FFT amplitude for the z component.
        """
        fig, ax = plt.subplots(3, 1, figsize=(10, 10))

        # Radial component (r)
        ax[0].plot(freq_r, fft_r, color='b')
        ax[0].set_title(f'FFT of Radial Component (r) for {name}')
        ax[0].set_xlabel('Frequency (Hz)')
        ax[0].set_ylabel('Amplitude')

        # Angular component (theta)
        ax[1].plot(freq_theta, fft_theta, color='g')
        ax[1].set_title(f'FFT of Angular Component (theta) for {name}')
        ax[1].set_xlabel('Frequency (Hz)')
        ax[1].set_ylabel('Amplitude')

        # z component
        ax[2].plot(freq_z, fft_z, color='r')
        ax[2].set_title(f'FFT of Z Component for {name}')
        ax[2].set_xlabel('Frequency (Hz)')
        ax[2].set_ylabel('Amplitude')

        plt.tight_layout()
        plt.show()

# Example usage of WaveFrequencyAnalyzer

# Create the dataset
filepath = 'SaturnModelDatabase/simulation_data/simulation 2025-01-11_00-13-15.bin'
dataset = Dataset(filepath)

# Initialize the analyzer with the dataset
analyzer = WaveFrequencyAnalyzer(dataset)

# Analyze a specific moon (e.g., "Daphnis")
analyzer.analyze_moon(moon_name="Daphnis")

# Analyze a specific test particle (e.g., particle 0)
analyzer.analyze_test_particles(particle_index=0)



{'r': (array([ 0.00000000e+00,  1.66500167e-06,  3.33000333e-06, ...,
         -4.99500500e-06, -3.33000333e-06, -1.66500167e-06]),
  array([1.31211380e+11, 7.99958337e+06, 8.12996036e+06, ...,
         8.51972326e+06, 8.12996036e+06, 7.99958337e+06])),
 'theta': (array([ 0.00000000e+00,  1.66500167e-06,  3.33000333e-06, ...,
         -4.99500500e-06, -3.33000333e-06, -1.66500167e-06]),
  array([3101.29776427,   30.00410198,   30.90164234, ...,   32.28033549,
           30.90164234,   30.00410198])),
 'z': (array([ 0.00000000e+00,  1.66500167e-06,  3.33000333e-06, ...,
         -4.99500500e-06, -3.33000333e-06, -1.66500167e-06]),
  array([1922627.86007745,  716069.70804339,  326688.55803931, ...,
          208022.3730623 ,  326688.55803931,  716069.70804339]))}

In [13]:
import numpy as np
from scipy.fft import fft, fftfreq

class FourierAnalysis:
    """
    Performs Fourier transform analysis on the dataset's test particles
    in cylindrical coordinates (r, theta, z) with respect to time.
    """
    def __init__(self, dataset):
        self.dataset = dataset  # This is the Dataset object
        self.num_timesteps = self.dataset.positions.shape[0]
        self.num_test_particles = self.dataset.num_test_particles
        self.dt = self.dataset.header['dt']  # Time step size from the header
        self.timesteps = np.linspace(0, self.num_timesteps * self.dt, self.num_timesteps)
        
        # Extract test particle positions (timesteps x test particles x 3D positions)
        self.test_particle_positions = self.dataset.positions[:, self.dataset.num_moons:, :3]

    def cartesian_to_cylindrical(self, x, y, z):
        """
        Converts Cartesian coordinates to cylindrical coordinates (r, theta, z).
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, theta, z

    def transform_to_cylindrical(self):
        """
        Converts the entire test particle dataset to cylindrical coordinates.
        Returns:
            ndarray: A 3D array of cylindrical coordinates (timesteps x test particles x (r, theta, z)).
        """
        r = np.sqrt(self.test_particle_positions[:, :, 0] ** 2 + self.test_particle_positions[:, :, 1] ** 2)
        theta = np.arctan2(self.test_particle_positions[:, :, 1], self.test_particle_positions[:, :, 0])
        z = self.test_particle_positions[:, :, 2]  # z remains the same

        return np.stack((r, theta, z), axis=-1)

    def compute_fourier_transform(self):
        """
        Computes the Fourier transform over time for all test particles in cylindrical coordinates.
        Returns:
            dict: A dictionary containing the Fourier transforms and frequencies for (r, theta, z).
        """
        # Convert the particle positions to cylindrical coordinates
        cylindrical_coords = self.transform_to_cylindrical()

        # Fourier transform each of the cylindrical coordinates
        fft_results = {}
        fft_results['r'] = fft(cylindrical_coords[:, :, 0], axis=0)  # Fourier transform for r
        fft_results['theta'] = fft(cylindrical_coords[:, :, 1], axis=0)  # Fourier transform for theta
        fft_results['z'] = fft(cylindrical_coords[:, :, 2], axis=0)  # Fourier transform for z

        # Compute the corresponding frequencies
        freqs = fftfreq(self.num_timesteps, d=self.dt)

        return {
            'frequencies': freqs,
            'fourier_r': fft_results['r'],
            'fourier_theta': fft_results['theta'],
            'fourier_z': fft_results['z']
        }

    def plot_fourier_transform(self, particle_index, max_frequency=None):
        """
        Plots the Fourier transform (amplitude spectrum) for a specific test particle.
        Args:
            particle_index (int): The index of the test particle to plot.
            max_frequency (float): Maximum frequency to display on the plot.
        """
        import matplotlib.pyplot as plt
        
        # Compute the Fourier transforms
        fourier_data = self.compute_fourier_transform()
        freqs = fourier_data['frequencies']
        
        # Filter frequencies if needed
        if max_frequency is not None:
            valid_indices = np.abs(freqs) <= max_frequency
        else:
            valid_indices = np.ones_like(freqs, dtype=bool)  # No filtering
        
        # Extract data for the specific particle
        fourier_r = np.abs(fourier_data['fourier_r'][:, particle_index])
        fourier_theta = np.abs(fourier_data['fourier_theta'][:, particle_index])
        fourier_z = np.abs(fourier_data['fourier_z'][:, particle_index])

        # Plot the amplitude spectra
        plt.figure(figsize=(14, 8))
        
        plt.subplot(311)
        plt.plot(freqs[valid_indices], fourier_r[valid_indices], label='Radial Distance (r)', color='r')
        plt.title(f"Fourier Transform of Particle {particle_index} in Cylindrical Coordinates")
        plt.xlabel('Frequency (Hz)')
        plt.ylabel('Amplitude')
        plt.grid(True)
        
        plt.subplot(312)
        plt.plot(freqs[valid_indices], fourier_theta[valid_indices], label='Azimuthal Angle (theta)', color='b')
        plt.xlabel('Frequency (Hz)')
        plt.ylabel('Amplitude')
        plt.grid(True)
        
        plt.subplot(313)
        plt.plot(freqs[valid_indices], fourier_z[valid_indices], label='Height (z)', color='g')
        plt.xlabel('Frequency (Hz)')
        plt.ylabel('Amplitude')
        plt.grid(True)

        plt.tight_layout()
        plt.show()

# Example usage:
fourier_analysis = FourierAnalysis(dataset)
fourier_analysis.plot_fourier_transform(particle_index=0, max_frequency=1e-5)


In [17]:
import numpy as np
from scipy.fft import fft, fftfreq

class CollectiveWaveFourier:
    def __init__(self, dataset, r_bins, theta_bins, z_bins):
        self.dataset = dataset
        self.num_timesteps = self.dataset.positions.shape[0]
        self.num_test_particles = self.dataset.num_test_particles
        self.dt = self.dataset.header['dt']  # Time step size from the header
        self.timesteps = np.linspace(0, self.num_timesteps * self.dt, self.num_timesteps)

        # Grid size for the cylindrical coordinates (r, theta, z)
        self.r_bins = r_bins
        self.theta_bins = theta_bins
        self.z_bins = z_bins

        # Extract test particle positions (timesteps x test particles x 3D positions)
        self.test_particle_positions = self.dataset.positions[:, self.dataset.num_moons:, :3]

    def cartesian_to_cylindrical(self, x, y, z):
        """
        Converts Cartesian coordinates to cylindrical coordinates (r, theta, z).
        """
        r = np.sqrt(x ** 2 + y ** 2)
        theta = np.arctan2(y, x)
        return r, theta, z

    def construct_wave_function(self):
        """
        Constructs a wave function representing the particle distribution
        in cylindrical coordinates (r, theta, z) over time.
        
        Returns:
            wave_function: A 4D array (timesteps x r_bins x theta_bins x z_bins) 
            representing the particle density in cylindrical coordinates over time.
        """
        # Initialize the wave function grid
        wave_function = np.zeros((self.num_timesteps, self.r_bins, self.theta_bins, self.z_bins))

        # Get the max values of r, theta, and z for grid scaling
        max_r = np.max(np.sqrt(self.test_particle_positions[:, :, 0]**2 + self.test_particle_positions[:, :, 1]**2))
        max_theta = 2 * np.pi  # Theta is always between -pi and pi
        max_z = np.max(self.test_particle_positions[:, :, 2]) - np.min(self.test_particle_positions[:, :, 2])

        # Define grid edges
        r_edges = np.linspace(0, max_r, self.r_bins + 1)
        theta_edges = np.linspace(-np.pi, np.pi, self.theta_bins + 1)
        z_edges = np.linspace(np.min(self.test_particle_positions[:, :, 2]), np.max(self.test_particle_positions[:, :, 2]), self.z_bins + 1)

        # Convert particle positions to cylindrical and bin them into the grid
        for t in range(self.num_timesteps):
            r, theta, z = self.cartesian_to_cylindrical(self.test_particle_positions[t, :, 0],
                                                        self.test_particle_positions[t, :, 1],
                                                        self.test_particle_positions[t, :, 2])
            # Digitize the positions into the grid
            r_indices = np.digitize(r, r_edges) - 1
            theta_indices = np.digitize(theta, theta_edges) - 1
            z_indices = np.digitize(z, z_edges) - 1

            # Count particles in each (r, theta, z) bin
            for i in range(self.num_test_particles):
                if 0 <= r_indices[i] < self.r_bins and 0 <= theta_indices[i] < self.theta_bins and 0 <= z_indices[i] < self.z_bins:
                    wave_function[t, r_indices[i], theta_indices[i], z_indices[i]] += 1

        return wave_function

    def perform_wave_fourier_transform(self):
        """
        Performs Fourier transform on the wave function over time.
        Returns:
            fourier_result: A 4D array (r_bins x theta_bins x z_bins x frequencies) 
            containing the Fourier transformed data for each spatial grid point.
            frequencies: The frequency array.
        """
        wave_function = self.construct_wave_function()

        # Fourier transform over the time axis (axis=0) for each (r, theta, z) point
        fourier_result = fft(wave_function, axis=0)
        
        # Compute corresponding frequencies
        frequencies = fftfreq(self.num_timesteps, d=self.dt)

        return fourier_result, frequencies

    def plot_wave_fourier_spectrum(self, r_index, theta_index, z_index, max_frequency=None):
        """
        Plots the Fourier transform spectrum for a specific (r, theta, z) grid point.
        """
        import matplotlib.pyplot as plt

        fourier_result, frequencies = self.perform_wave_fourier_transform()

        # Extract the Fourier data for the specific grid point
        wave_spectrum = np.abs(fourier_result[:, r_index, theta_index, z_index])

        # Filter frequencies if necessary
        if max_frequency is not None:
            valid_indices = np.abs(frequencies) <= max_frequency
        else:
            valid_indices = np.ones_like(frequencies, dtype=bool)

        # Plot the spectrum
        plt.figure(figsize=(10, 6))
        plt.plot(frequencies[valid_indices], wave_spectrum[valid_indices])
        plt.title(f"Fourier Transform of Wave Function at (r_bin={r_index}, theta_bin={theta_index}, z_bin={z_index})")
        plt.xlabel('Frequency (Hz)')
        plt.ylabel('Amplitude')
        plt.grid(True)
        plt.show()

# Example usage:
collective_wave_fourier = CollectiveWaveFourier(dataset, r_bins=20, theta_bins=10, z_bins=20)
collective_wave_fourier.plot_wave_fourier_spectrum(r_index=10, theta_index=4, z_index=10, max_frequency=1e-7)


In [19]:
import numpy as np
from scipy.fft import fft, fftfreq
import matplotlib.pyplot as plt

# Mock data representing the particle positions over time (timesteps x particles x (x, y, z))
# Here, I'm generating random data for simplicity, but you'd load your actual data.
timesteps = 100
particles = 1000
positions = np.random.rand(timesteps, particles, 3) * 10  # Replace with your data

# Define grid resolution (bins in r, theta, z)
r_bins = 50
theta_bins = 50
z_bins = 50

# Define a maximum range for cylindrical coordinates (assuming all particles stay within these ranges)
max_r = 10
max_theta = 2 * np.pi
max_z = 10

# Time step size
dt = 0.1  # Change based on your actual time step size
time = np.linspace(0, timesteps * dt, timesteps)

# Function to convert Cartesian to cylindrical coordinates
def cartesian_to_cylindrical(x, y, z):
    r = np.sqrt(x**2 + y**2)
    theta = np.arctan2(y, x)
    return r, theta, z

# Initialize the wave function grid (timesteps x r_bins x theta_bins x z_bins)
wave_function = np.zeros((timesteps, r_bins, theta_bins, z_bins))

# Define the edges of the bins for r, theta, and z
r_edges = np.linspace(0, max_r, r_bins + 1)
theta_edges = np.linspace(-np.pi, np.pi, theta_bins + 1)
z_edges = np.linspace(0, max_z, z_bins + 1)

# Loop over time steps to fill the wave function grid
for t in range(timesteps):
    # Get the particle positions at time t
    x = positions[t, :, 0]
    y = positions[t, :, 1]
    z = positions[t, :, 2]

    # Convert to cylindrical coordinates
    r, theta, z = cartesian_to_cylindrical(x, y, z)

    # Digitize the positions into the bins
    r_indices = np.digitize(r, r_edges) - 1
    theta_indices = np.digitize(theta, theta_edges) - 1
    z_indices = np.digitize(z, z_edges) - 1

    # Count particles in each (r, theta, z) bin
    for i in range(particles):
        if 0 <= r_indices[i] < r_bins and 0 <= theta_indices[i] < theta_bins and 0 <= z_indices[i] < z_bins:
            wave_function[t, r_indices[i], theta_indices[i], z_indices[i]] += 1

# Now we have a "wave function" that represents the particle distribution over time in cylindrical coordinates

# Let's Fourier transform this wave function over time for each spatial location
fourier_transform = np.zeros((r_bins, theta_bins, z_bins, timesteps), dtype=complex)
frequencies = fftfreq(timesteps, d=dt)

# Fourier transform over time for each (r, theta, z) bin
for r_idx in range(r_bins):
    for theta_idx in range(theta_bins):
        for z_idx in range(z_bins):
            fourier_transform[r_idx, theta_idx, z_idx, :] = fft(wave_function[:, r_idx, theta_idx, z_idx])

# Example: Plot the Fourier spectrum at a specific (r, theta, z) bin
r_bin = 25
theta_bin = 25
z_bin = 25

# Get the Fourier amplitude for this spatial bin
spectrum = np.abs(fourier_transform[r_bin, theta_bin, z_bin, :])

# Plot the spectrum
plt.plot(frequencies, spectrum)
plt.title(f"Fourier Spectrum at (r={r_bin}, theta={theta_bin}, z={z_bin})")
plt.xlabel("Frequency")
plt.ylabel("Amplitude")
plt.grid(True)
plt.show()


In [22]:
import numpy as np
from scipy.fft import fft, fftfreq
import matplotlib.pyplot as plt

# Mock data representing structured particle motion (instead of random)
timesteps = 100
particles = 1000
positions = np.zeros((timesteps, particles, 3))  # Timesteps x Particles x (x, y, z)

# Time step size
dt = 0.1
time = np.linspace(0, timesteps * dt, timesteps)

# Create a structured, periodic motion for the particles
for t in range(timesteps):
    # For simplicity, let's create circular motion in the xy-plane and sinusoidal motion in z
    theta_motion = 2 * np.pi * t / timesteps
    r = 5 + 2 * np.sin(2 * np.pi * t / timesteps)  # Oscillating radius
    theta = np.linspace(0, 2 * np.pi, particles) + theta_motion  # Rotate particles around z-axis
    z = 5 + 2 * np.cos(2 * np.pi * t / timesteps)  # Oscillating z-motion
    
    # Convert cylindrical to Cartesian
    x = r * np.cos(theta)
    y = r * np.sin(theta)
    
    positions[t, :, 0] = x
    positions[t, :, 1] = y
    positions[t, :, 2] = z

# Define grid resolution (bins in r and z)
r_bins = 50
z_bins = 50

# Define a maximum range for cylindrical coordinates (assuming all particles stay within these ranges)
max_r = 10
max_z = 10

# Function to convert Cartesian to cylindrical coordinates
def cartesian_to_cylindrical(x, y, z):
    r = np.sqrt(x**2 + y**2)
    return r, z

# Initialize the wave function grid (timesteps x r_bins x z_bins)
wave_function = np.zeros((timesteps, r_bins, z_bins))

# Define the edges of the bins for r and z
r_edges = np.linspace(0, max_r, r_bins + 1)
z_edges = np.linspace(0, max_z, z_bins + 1)

# Loop over time steps to fill the wave function grid
for t in range(timesteps):
    # Get the particle positions at time t
    x = positions[t, :, 0]
    y = positions[t, :, 1]
    z = positions[t, :, 2]

    # Convert to cylindrical coordinates
    r, z = cartesian_to_cylindrical(x, y, z)

    # Digitize the positions into the bins
    r_indices = np.digitize(r, r_edges) - 1
    z_indices = np.digitize(z, z_edges) - 1

    # Count particles in each (r, z) bin
    for i in range(particles):
        if 0 <= r_indices[i] < r_bins and 0 <= z_indices[i] < z_bins:
            wave_function[t, r_indices[i], z_indices[i]] += 1

# Now we have a "wave function" that represents the particle distribution over time in cylindrical coordinates

# Let's create a 2D visualization of the wave function at a few different time steps
fig, ax = plt.subplots(1, 3, figsize=(18, 5))

for i, t in enumerate([0, timesteps//2, timesteps-1]):
    im = ax[i].imshow(wave_function[t, :, :], extent=[0, max_z, 0, max_r], origin='lower', aspect='auto', cmap='viridis')
    ax[i].set_title(f"Particle Density at Time {t*dt:.2f}")
    ax[i].set_xlabel("z")
    ax[i].set_ylabel("r")
    fig.colorbar(im, ax=ax[i])

plt.tight_layout()
plt.show()

# Let's Fourier transform this wave function over time for each (r, z) bin
fourier_transform = np.zeros((r_bins, z_bins, timesteps), dtype=complex)
frequencies = fftfreq(timesteps, d=dt)

# Fourier transform over time for each (r, z) bin
for r_idx in range(r_bins):
    for z_idx in range(z_bins):
        fourier_transform[r_idx, z_idx, :] = fft(wave_function[:, r_idx, z_idx])

# Plot the Fourier spectrum for a specific (r, z) bin
r_bin = 25  # Mid-point in r
z_bin = 25  # Mid-point in z

# Get the Fourier amplitude for this spatial bin
spectrum = np.abs(fourier_transform[r_bin, z_bin, :])

# Plot the spectrum
plt.plot(frequencies, spectrum)
plt.title(f"Fourier Spectrum at (r={r_bin}, z={z_bin})")
plt.xlabel("Frequency")
plt.ylabel("Amplitude")
plt.grid(True)
plt.show()

# Visualize the Fourier transform for all (r, z) bins at a single frequency component (e.g., the first harmonic)
frequency_bin = 5  # Choose a specific frequency component to visualize
fourier_amplitude_at_frequency = np.abs(fourier_transform[:, :, frequency_bin])

# 2D plot of the Fourier amplitude in (r, z) space
plt.imshow(fourier_amplitude_at_frequency, extent=[0, max_z, 0, max_r], origin='lower', aspect='auto', cmap='plasma')
plt.title(f"Fourier Amplitude at Frequency {frequencies[frequency_bin]:.2f}")
plt.xlabel("z")
plt.ylabel("r")
plt.colorbar(label="Amplitude")
plt.show()


In [21]:
from mpl_toolkits.mplot3d import Axes3D

# Choose a time step to visualize the particle positions
time_step = 50  # Pick any time step to visualize

# Get the particle positions at the chosen time step
x = positions[time_step, :, 0]
y = positions[time_step, :, 1]
z = positions[time_step, :, 2]

# 3D Scatter plot of the particles
fig = plt.figure(figsize=(12, 6))

ax = fig.add_subplot(121, projection='3d')
ax.scatter(x, y, z, c='b', marker='o', s=10)
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')
ax.set_title(f'3D Particle Positions at Time Step {time_step}')

# 2D scatter plot in the r-z plane (cylindrical coordinates)
r, z_cyl = cartesian_to_cylindrical(x, y, z)
ax2 = fig.add_subplot(122)
ax2.scatter(r, z_cyl, c='r', s=10)
ax2.set_xlabel('r')
ax2.set_ylabel('z')
ax2.set_title(f'2D Projection in r-z Plane at Time Step {time_step}')

plt.tight_layout()
plt.show()

# 2D scatter plot in the x-y plane (top-down view)
plt.scatter(x, y, c='g', s=10)
plt.xlabel('x')
plt.ylabel('y')
plt.title(f'2D Projection in x-y Plane at Time Step {time_step}')
plt.axis('equal')  # Ensure equal scaling on both axes
plt.grid(True)
plt.show()


In [13]:
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as anim
from mpl_toolkits.mplot3d import Axes3D

# Parameters
x_max, y_max = 30, 10  # x_max is now 3 times larger (3 * original x_max)
num_points = 10000  # Number of particles
frames = 100
wave_length = 2 * np.pi
speed = 0.2
amplitude_1 = 0.4  # Amplitude for the first cosine wave
amplitude_2 = 0.3  # Amplitude for the second cosine wave
amplitude_3 = 0.2  # Amplitude for the third cosine wave
num_frequencies = 3  # Number of frequencies in the sum
length_of_x = 1.0  # Controls the "stretch" along the x-axis
y_translation_factor = 0.1  # Controls how much the y-translation happens for each wave

# Define different frequencies for the sum of cosines
frequencies = [1, 2, 3]  # Frequencies of the cosines
phases = np.random.uniform(0, 2 * np.pi, num_frequencies)  # Random phase shifts for each frequency

# Create random initial positions for x, y
x = np.random.uniform(0, x_max, num_points)
y = np.random.uniform(0, y_max, num_points)

# Set up the figure and axis
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.set_xlim(0, x_max)
ax.set_ylim(0, y_max)
ax.set_zlim(-1, 1)

# Initialize scatter plot with empty data
scat = ax.scatter([], [], [], s=5)  # 's' controls the size of the points

def update_wave(frame):
    # Create a sum of cosines with different frequencies and random phases
    z = np.zeros_like(x)
    for i in range(num_frequencies):
        # Introduce a y-dependent phase shift for a slight translation effect in y
        y_shift = y_translation_factor * y * i  # Different shifts for each wave
        z += (amplitude_1 if i == 0 else amplitude_2 if i == 1 else amplitude_3) * \
            np.cos(frequencies[i] * 2 * np.pi * (x - speed * frame) / wave_length + phases[i] + y_shift)

    # Apply length_of_x scaling factor to control the "stretch" of the wave
    z *= length_of_x

    # Update the positions and z-values for scatter plot
    scat._offsets3d = (x, y, z)

    return scat,

# Create the animation
ani = anim.FuncAnimation(fig, update_wave, frames=frames, interval=50, blit=False)

plt.show()


In [6]:
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as anim
from mpl_toolkits.mplot3d import Axes3D

# Parameters
x_max, y_max = 30, 10  # x_max is now 3 times larger (3 * original x_max)
num_points = 1000  # Number of particles
frames = 100
wave_length = 2 * np.pi
speed = 0.2
amplitude = 1.0  # Amplitude of the wave (adjust this value)
num_frequencies = 3  # Number of frequencies in the sum
length_of_x = 1.0  # Controls the "stretch" along the x-axis

# Define different frequencies for the sum of cosines
frequencies = [1, 2, 3]  # Frequencies of the cosines
phases = np.random.uniform(0, 2 * np.pi, num_frequencies)  # Random phase shifts for each frequency

# Create a meshgrid for x, y values (create a 2D grid of points instead of random positions)
x = np.linspace(0, x_max, num_points)
y = np.linspace(0, y_max, num_points)
X, Y = np.meshgrid(x, y)  # Create a mesh grid from the x, y values

# Set up the figure and axis for both the wave and the Fourier transform
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 6))

# Set up the 3D plot for the wave
ax1 = fig.add_subplot(121, projection='3d')
ax1.set_xlim(0, x_max)
ax1.set_ylim(0, y_max)
ax1.set_zlim(-1, 1)
scat = ax1.scatter([], [], [], s=5)  # 's' controls the size of the points

# Set up the plot for the 2D Fourier transform
ax2.set_xlim(0, x_max)  # X-axis for frequencies in x-direction
ax2.set_ylim(0, y_max)  # Y-axis for frequencies in y-direction
img = ax2.imshow(np.zeros((num_points, num_points)), cmap='viridis', origin='lower', animated=True)

def update_wave(frame):
    # Create a sum of cosines with different frequencies and random phases
    z = np.zeros_like(X)
    for i in range(num_frequencies):
        z += amplitude * np.cos(frequencies[i] * 2 * np.pi * (X - speed * frame) / wave_length + phases[i])

    # Apply length_of_x scaling factor to control the "stretch" of the wave
    z *= length_of_x

    # Update the positions and z-values for scatter plot (3D wave)
    scat._offsets3d = (X.flatten(), Y.flatten(), z.flatten())

    # Compute the 2D Fourier Transform (FFT2) of the wave
    z_fft2 = np.fft.fft2(z)  # 2D FFT of the signal
    z_fft2_magnitude = np.abs(z_fft2)  # Magnitude of the Fourier coefficients

    # Update the Fourier transform plot
    img.set_array(np.log(z_fft2_magnitude + 1))  # Use log scale for better visualization

    return scat, img,

# Create the animation
ani = anim.FuncAnimation(fig, update_wave, frames=frames, interval=50, blit=False)

plt.show()
